File: kivg/rendering/text_renderer.py
"""
Text rendering functionality for OpenCV-based SVG rendering.
Handles drawing text elements with animation support.
"""
import cv2
import numpy as np
from typing import List, Tuple, Dict, Any, Optional

from ..core.canvas import OpenCVCanvas


class TextRenderer:
    """Handles rendering of SVG text elements to OpenCV canvas."""
    
    # Map SVG font weights to OpenCV font thickness adjustments
    FONT_WEIGHT_MAP = {
        'normal': 1,
        'bold': 2,
        '100': 1,
        '200': 1,
        '300': 1,
        '400': 1,
        '500': 1,
        '600': 2,
        '700': 2,
        '800': 2,
        '900': 3
    }
    
    # Available OpenCV fonts (limited selection)
    OPENCV_FONTS = {
        'sans-serif': cv2.FONT_HERSHEY_SIMPLEX,
        'serif': cv2.FONT_HERSHEY_TRIPLEX,
        'monospace': cv2.FONT_HERSHEY_PLAIN,
        'arial': cv2.FONT_HERSHEY_SIMPLEX,
        'helvetica': cv2.FONT_HERSHEY_SIMPLEX,
        'times': cv2.FONT_HERSHEY_TRIPLEX,
        'times new roman': cv2.FONT_HERSHEY_TRIPLEX,
        'courier': cv2.FONT_HERSHEY_PLAIN,
        'courier new': cv2.FONT_HERSHEY_PLAIN
    }
    
    @staticmethod
    def get_opencv_font(font_family: str, font_style: str = 'normal') -> int:
        """
        Map SVG font family to OpenCV font.
        
        Args:
            font_family: SVG font family name
            font_style: Font style (normal, italic)
            
        Returns:
            OpenCV font constant
        """
        font_family_lower = font_family.lower().strip()
        
        # Try to find matching font
        for font_name, opencv_font in TextRenderer.OPENCV_FONTS.items():
            if font_name in font_family_lower:
                # Add italic variant if requested
                if font_style == 'italic' or font_style == 'oblique':
                    return opencv_font | cv2.FONT_ITALIC
                return opencv_font
        
        # Default to sans-serif
        base_font = cv2.FONT_HERSHEY_SIMPLEX
        if font_style == 'italic' or font_style == 'oblique':
            return base_font | cv2.FONT_ITALIC
        return base_font
    
    @staticmethod
    def calculate_font_scale(font_size: float) -> float:
        """
        Calculate OpenCV font scale from SVG font size.
        
        Args:
            font_size: SVG font size in pixels
            
        Returns:
            OpenCV font scale value
        """
        # OpenCV font scale is approximately: pixel_height = 30 * scale for HERSHEY_SIMPLEX
        # So scale = font_size / 30
        return font_size / 30.0
    
    @staticmethod
    def draw_text(canvas: OpenCVCanvas, text_data: Dict[str, Any],
                  scale_x: float = 1.0, scale_y: float = 1.0,
                  offset_x: float = 0, offset_y: float = 0,
                  opacity: float = 1.0,
                  char_reveal: int = -1) -> None:
        """
        Draw text element on the canvas.
        
        Args:
            canvas: OpenCVCanvas to draw on
            text_data: Dictionary with text properties
            scale_x: X scale factor for coordinate transformation
            scale_y: Y scale factor for coordinate transformation
            offset_x: X offset for positioning
            offset_y: Y offset for positioning
            opacity: Opacity multiplier (0-1)
            char_reveal: Number of characters to reveal (-1 for all)
        """
        text = text_data['text']
        if char_reveal >= 0:
            text = text[:char_reveal]
        
        if not text:
            return
        
        # Transform coordinates
        x = int(text_data['x'] * scale_x + offset_x)
        y = int(text_data['y'] * scale_y + offset_y)
        
        # Get font settings
        font = TextRenderer.get_opencv_font(
            text_data.get('font_family', 'sans-serif'),
            text_data.get('font_style', 'normal')
        )
        font_size = text_data.get('font_size', 16)
        font_scale = TextRenderer.calculate_font_scale(font_size * min(scale_x, scale_y))
        
        # Get font weight/thickness
        font_weight = text_data.get('font_weight', 'normal')
        thickness = TextRenderer.FONT_WEIGHT_MAP.get(str(font_weight), 1)
        
        # Calculate text size for alignment
        (text_width, text_height), baseline = cv2.getTextSize(
            text, font, font_scale, thickness
        )
        
        # Adjust position based on text-anchor
        text_anchor = text_data.get('text_anchor', 'start')
        if text_anchor == 'middle':
            x -= text_width // 2
        elif text_anchor == 'end':
            x -= text_width
        
        # Adjust baseline
        dominant_baseline = text_data.get('dominant_baseline', 'auto')
        if dominant_baseline in ('middle', 'central'):
            y += text_height // 2
        elif dominant_baseline in ('hanging', 'text-before-edge'):
            y += text_height
        
        # Get colors
        fill_color = text_data.get('fill', [0, 0, 0, 1])
        stroke_color = text_data.get('stroke')
        stroke_width = text_data.get('stroke_width', 1)
        text_opacity = text_data.get('opacity', 1.0) * opacity
        
        # Apply letter spacing if specified
        letter_spacing = text_data.get('letter_spacing', 0) * scale_x
        
        # Convert colors to 0-255 range
        if fill_color:
            fill_rgba = tuple(int(c * 255) for c in fill_color[:3])
            fill_rgba = fill_rgba + (int(fill_color[3] * 255 * text_opacity),)
        else:
            fill_rgba = None
        
        if stroke_color:
            stroke_rgba = tuple(int(c * 255) for c in stroke_color[:3])
            stroke_rgba = stroke_rgba + (int(stroke_color[3] * 255 * text_opacity),)
        else:
            stroke_rgba = None
        
        # Draw text with letter spacing
        if letter_spacing > 0:
            TextRenderer._draw_text_with_spacing(
                canvas, text, x, y, font, font_scale, thickness,
                letter_spacing, fill_rgba, stroke_rgba, stroke_width,
                text_data.get('text_decoration', 'none'),
                text_height
            )
        else:
            TextRenderer._draw_text_simple(
                canvas, text, x, y, font, font_scale, thickness,
                fill_rgba, stroke_rgba, stroke_width,
                text_data.get('text_decoration', 'none'),
                text_width, text_height
            )
    
    @staticmethod
    def _draw_text_simple(canvas: OpenCVCanvas, text: str,
                          x: int, y: int, font: int, font_scale: float,
                          thickness: int, fill_rgba: Optional[Tuple],
                          stroke_rgba: Optional[Tuple], stroke_width: float,
                          text_decoration: str, text_width: int,
                          text_height: int) -> None:
        """Draw text without letter spacing."""
        # Draw stroke first (if any)
        if stroke_rgba and stroke_width:
            stroke_thickness = thickness + int(stroke_width * 2)
            if stroke_rgba[3] > 0:
                cv2.putText(
                    canvas.image, text, (x, y),
                    font, font_scale, stroke_rgba[:3],
                    stroke_thickness, cv2.LINE_AA
                )
        
        # Draw fill
        if fill_rgba and fill_rgba[3] > 0:
            cv2.putText(
                canvas.image, text, (x, y),
                font, font_scale, fill_rgba[:3],
                thickness, cv2.LINE_AA
            )
        
        # Draw text decoration
        TextRenderer._draw_decoration(
            canvas, text_decoration, x, y, text_width, text_height,
            fill_rgba, thickness
        )
    
    @staticmethod
    def _draw_text_with_spacing(canvas: OpenCVCanvas, text: str,
                                 x: int, y: int, font: int, font_scale: float,
                                 thickness: int, letter_spacing: float,
                                 fill_rgba: Optional[Tuple],
                                 stroke_rgba: Optional[Tuple],
                                 stroke_width: float, text_decoration: str,
                                 text_height: int) -> None:
        """Draw text with custom letter spacing."""
        current_x = x
        total_width = 0
        
        for char in text:
            # Get character width
            (char_width, _), _ = cv2.getTextSize(char, font, font_scale, thickness)
            
            # Draw stroke first
            if stroke_rgba and stroke_width:
                stroke_thickness = thickness + int(stroke_width * 2)
                if stroke_rgba[3] > 0:
                    cv2.putText(
                        canvas.image, char, (int(current_x), y),
                        font, font_scale, stroke_rgba[:3],
                        stroke_thickness, cv2.LINE_AA
                    )
            
            # Draw fill
            if fill_rgba and fill_rgba[3] > 0:
                cv2.putText(
                    canvas.image, char, (int(current_x), y),
                    font, font_scale, fill_rgba[:3],
                    thickness, cv2.LINE_AA
                )
            
            current_x += char_width + letter_spacing
            total_width += char_width + letter_spacing
        
        # Draw text decoration
        TextRenderer._draw_decoration(
            canvas, text_decoration, x, y, int(total_width - letter_spacing),
            text_height, fill_rgba, thickness
        )
    
    @staticmethod
    def _draw_decoration(canvas: OpenCVCanvas, decoration: str,
                         x: int, y: int, text_width: int, text_height: int,
                         color: Optional[Tuple], thickness: int) -> None:
        """Draw text decoration (underline, strikethrough, etc.)."""
        if not color or color[3] == 0 or decoration == 'none':
            return
        
        line_thickness = max(1, thickness // 2)
        
        if 'underline' in decoration:
            line_y = y + text_height // 4
            cv2.line(canvas.image, (x, line_y), (x + text_width, line_y),
                    color[:3], line_thickness, cv2.LINE_AA)
        
        if 'line-through' in decoration or 'strikethrough' in decoration:
            line_y = y - text_height // 3
            cv2.line(canvas.image, (x, line_y), (x + text_width, line_y),
                    color[:3], line_thickness, cv2.LINE_AA)
        
        if 'overline' in decoration:
            line_y = y - text_height
            cv2.line(canvas.image, (x, line_y), (x + text_width, line_y),
                    color[:3], line_thickness, cv2.LINE_AA)
    
    @staticmethod
    def get_text_bounds(text_data: Dict[str, Any],
                        scale_x: float = 1.0, scale_y: float = 1.0) -> Tuple[int, int, int, int]:
        """
        Calculate bounding box for text element.
        
        Args:
            text_data: Dictionary with text properties
            scale_x: X scale factor
            scale_y: Y scale factor
            
        Returns:
            Tuple (x, y, width, height)
        """
        text = text_data['text']
        font = TextRenderer.get_opencv_font(
            text_data.get('font_family', 'sans-serif'),
            text_data.get('font_style', 'normal')
        )
        font_size = text_data.get('font_size', 16)
        font_scale = TextRenderer.calculate_font_scale(font_size * min(scale_x, scale_y))
        
        font_weight = text_data.get('font_weight', 'normal')
        thickness = TextRenderer.FONT_WEIGHT_MAP.get(str(font_weight), 1)
        
        (text_width, text_height), baseline = cv2.getTextSize(
            text, font, font_scale, thickness
        )
        
        x = int(text_data['x'] * scale_x)
        y = int(text_data['y'] * scale_y)
        
        return (x, y - text_height, text_width, text_height + baseline)
    
    @staticmethod
    def get_char_position(text_data: Dict[str, Any], char_index: int,
                          scale_x: float = 1.0, scale_y: float = 1.0,
                          offset_x: float = 0, offset_y: float = 0) -> Optional[Tuple[int, int]]:
        """
        Get the position (x, y) of a specific character in the text.
        
        Args:
            text_data: Dictionary with text properties
            char_index: Index of the character (0-based)
            scale_x: X scale factor
            scale_y: Y scale factor
            offset_x: X offset for positioning
            offset_y: Y offset for positioning
            
        Returns:
            Tuple (x, y) of the character position, or None if invalid
        """
        text = text_data['text']
        if char_index < 0 or char_index > len(text):
            return None
        
        if char_index == 0:
            # Return the starting position
            x = int(text_data['x'] * scale_x + offset_x)
            y = int(text_data['y'] * scale_y + offset_y)
            return (x, y)
        
        # Get font settings
        font = TextRenderer.get_opencv_font(
            text_data.get('font_family', 'sans-serif'),
            text_data.get('font_style', 'normal')
        )
        font_size = text_data.get('font_size', 16)
        font_scale = TextRenderer.calculate_font_scale(font_size * min(scale_x, scale_y))
        
        font_weight = text_data.get('font_weight', 'normal')
        thickness = TextRenderer.FONT_WEIGHT_MAP.get(str(font_weight), 1)
        
        # Calculate base position
        x = int(text_data['x'] * scale_x + offset_x)
        y = int(text_data['y'] * scale_y + offset_y)
        
        # Calculate full text size for alignment
        full_text = text_data['text']
        (full_width, text_height), _ = cv2.getTextSize(
            full_text, font, font_scale, thickness
        )
        
        # Adjust position based on text-anchor
        text_anchor = text_data.get('text_anchor', 'start')
        if text_anchor == 'middle':
            x -= full_width // 2
        elif text_anchor == 'end':
            x -= full_width
        
        # Adjust baseline
        dominant_baseline = text_data.get('dominant_baseline', 'auto')
        if dominant_baseline in ('middle', 'central'):
            y += text_height // 2
        elif dominant_baseline in ('hanging', 'text-before-edge'):
            y += text_height
        
        # Calculate width of text up to char_index
        letter_spacing = text_data.get('letter_spacing', 0) * scale_x
        
        if letter_spacing > 0:
            # Calculate width character by character
            current_x = x
            for i in range(min(char_index, len(text))):
                char = text[i]
                (char_width, _), _ = cv2.getTextSize(char, font, font_scale, thickness)
                current_x += char_width + letter_spacing
            x = int(current_x)
        else:
            # Calculate width of substring
            substring = text[:char_index]
            (sub_width, _), _ = cv2.getTextSize(substring, font, font_scale, thickness)
            x += sub_width
        
        return (x, y)

File: migration/rendering/TextRenderer.js
/**
 * Text rendering functionality for OpenCV.js-based SVG rendering.
 * Handles drawing text elements with animation support.
 * 
 * This is the JavaScript equivalent of kivg/rendering/text_renderer.py
 */

/**
 * Handles rendering of SVG text elements to OpenCV canvas.
 */
class TextRenderer {
    /**
     * Map SVG font weights to OpenCV font thickness adjustments.
     */
    static FONT_WEIGHT_MAP = {
        'normal': 1,
        'bold': 2,
        '100': 1,
        '200': 1,
        '300': 1,
        '400': 1,
        '500': 1,
        '600': 2,
        '700': 2,
        '800': 2,
        '900': 3
    };

    /**
     * Available OpenCV fonts (limited selection).
     */
    static OPENCV_FONTS = {
        'sans-serif': () => cv.FONT_HERSHEY_SIMPLEX,
        'serif': () => cv.FONT_HERSHEY_TRIPLEX,
        'monospace': () => cv.FONT_HERSHEY_PLAIN,
        'arial': () => cv.FONT_HERSHEY_SIMPLEX,
        'helvetica': () => cv.FONT_HERSHEY_SIMPLEX,
        'times': () => cv.FONT_HERSHEY_TRIPLEX,
        'times new roman': () => cv.FONT_HERSHEY_TRIPLEX,
        'courier': () => cv.FONT_HERSHEY_PLAIN,
        'courier new': () => cv.FONT_HERSHEY_PLAIN
    };

    /**
     * Map SVG font family to OpenCV font.
     * @param {string} fontFamily - SVG font family name
     * @param {string} fontStyle - Font style (normal, italic)
     * @returns {number} OpenCV font constant
     */
    static getOpenCVFont(fontFamily, fontStyle = 'normal') {
        const fontFamilyLower = fontFamily.toLowerCase().trim();

        // Try to find matching font
        for (const [fontName, fontGetter] of Object.entries(TextRenderer.OPENCV_FONTS)) {
            if (fontFamilyLower.includes(fontName)) {
                let font = fontGetter();
                // Add italic variant if requested
                if (fontStyle === 'italic' || fontStyle === 'oblique') {
                    font = font | cv.FONT_ITALIC;
                }
                return font;
            }
        }

        // Default to sans-serif
        let baseFont = cv.FONT_HERSHEY_SIMPLEX;
        if (fontStyle === 'italic' || fontStyle === 'oblique') {
            baseFont = baseFont | cv.FONT_ITALIC;
        }
        return baseFont;
    }

    /**
     * Calculate OpenCV font scale from SVG font size.
     * @param {number} fontSize - SVG font size in pixels
     * @returns {number} OpenCV font scale value
     */
    static calculateFontScale(fontSize) {
        // OpenCV font scale is approximately: pixel_height = 30 * scale for HERSHEY_SIMPLEX
        return fontSize / 30.0;
    }

    /**
     * Draw text element on the canvas.
     * @param {OpenCVCanvas} canvas - OpenCVCanvas to draw on
     * @param {Object} textData - Dictionary with text properties
     * @param {number} scaleX - X scale factor for coordinate transformation
     * @param {number} scaleY - Y scale factor for coordinate transformation
     * @param {number} offsetX - X offset for positioning
     * @param {number} offsetY - Y offset for positioning
     * @param {number} opacity - Opacity multiplier (0-1)
     * @param {number} charReveal - Number of characters to reveal (-1 for all)
     */
    static drawText(canvas, textData, scaleX = 1.0, scaleY = 1.0, offsetX = 0, offsetY = 0, opacity = 1.0, charReveal = -1) {
        let text = textData.text;
        if (charReveal >= 0) {
            text = text.substring(0, charReveal);
        }

        if (!text) {
            return;
        }

        // Transform coordinates
        let x = Math.floor(textData.x * scaleX + offsetX);
        let y = Math.floor(textData.y * scaleY + offsetY);

        // Get font settings
        const font = TextRenderer.getOpenCVFont(
            textData.font_family || 'sans-serif',
            textData.font_style || 'normal'
        );
        const fontSize = textData.font_size || 16;
        const fontScale = TextRenderer.calculateFontScale(fontSize * Math.min(scaleX, scaleY));

        // Get font weight/thickness
        const fontWeight = textData.font_weight || 'normal';
        const thickness = TextRenderer.FONT_WEIGHT_MAP[String(fontWeight)] || 1;

        // Calculate text size for alignment
        const textSize = cv.getTextSize(text, font, fontScale, thickness);
        const textWidth = textSize.width;
        const textHeight = textSize.height;

        // Adjust position based on text-anchor
        const textAnchor = textData.text_anchor || 'start';
        if (textAnchor === 'middle') {
            x -= Math.floor(textWidth / 2);
        } else if (textAnchor === 'end') {
            x -= textWidth;
        }

        // Adjust baseline
        const dominantBaseline = textData.dominant_baseline || 'auto';
        if (['middle', 'central'].includes(dominantBaseline)) {
            y += Math.floor(textHeight / 2);
        } else if (['hanging', 'text-before-edge'].includes(dominantBaseline)) {
            y += textHeight;
        }

        // Get colors
        const fillColor = textData.fill || [0, 0, 0, 1];
        const strokeColor = textData.stroke;
        const strokeWidth = textData.stroke_width || 1;
        const textOpacity = (textData.opacity || 1.0) * opacity;

        // Apply letter spacing if specified
        const letterSpacing = (textData.letter_spacing || 0) * scaleX;

        // Convert colors to 0-255 range
        let fillRgba = null;
        if (fillColor) {
            fillRgba = [
                Math.floor(fillColor[0] * 255),
                Math.floor(fillColor[1] * 255),
                Math.floor(fillColor[2] * 255),
                Math.floor(fillColor[3] * 255 * textOpacity)
            ];
        }

        let strokeRgba = null;
        if (strokeColor) {
            strokeRgba = [
                Math.floor(strokeColor[0] * 255),
                Math.floor(strokeColor[1] * 255),
                Math.floor(strokeColor[2] * 255),
                Math.floor(strokeColor[3] * 255 * textOpacity)
            ];
        }

        // Draw text with letter spacing
        if (letterSpacing > 0) {
            TextRenderer._drawTextWithSpacing(
                canvas, text, x, y, font, fontScale, thickness,
                letterSpacing, fillRgba, strokeRgba, strokeWidth,
                textData.text_decoration || 'none',
                textHeight
            );
        } else {
            TextRenderer._drawTextSimple(
                canvas, text, x, y, font, fontScale, thickness,
                fillRgba, strokeRgba, strokeWidth,
                textData.text_decoration || 'none',
                textWidth, textHeight
            );
        }
    }

    /**
     * Draw text without letter spacing.
     * @private
     */
    static _drawTextSimple(canvas, text, x, y, font, fontScale, thickness, fillRgba, strokeRgba, strokeWidth, textDecoration, textWidth, textHeight) {
        // Draw stroke first (if any)
        if (strokeRgba && strokeWidth) {
            const strokeThickness = thickness + Math.floor(strokeWidth * 2);
            if (strokeRgba[3] > 0) {
                cv.putText(
                    canvas.mat,
                    text,
                    new cv.Point(x, y),
                    font,
                    fontScale,
                    new cv.Scalar(strokeRgba[0], strokeRgba[1], strokeRgba[2], 255),
                    strokeThickness,
                    cv.LINE_AA
                );
            }
        }

        // Draw fill
        if (fillRgba && fillRgba[3] > 0) {
            cv.putText(
                canvas.mat,
                text,
                new cv.Point(x, y),
                font,
                fontScale,
                new cv.Scalar(fillRgba[0], fillRgba[1], fillRgba[2], 255),
                thickness,
                cv.LINE_AA
            );
        }

        // Draw text decoration
        TextRenderer._drawDecoration(
            canvas, textDecoration, x, y, textWidth, textHeight,
            fillRgba, thickness
        );
    }

    /**
     * Draw text with custom letter spacing.
     * @private
     */
    static _drawTextWithSpacing(canvas, text, x, y, font, fontScale, thickness, letterSpacing, fillRgba, strokeRgba, strokeWidth, textDecoration, textHeight) {
        let currentX = x;
        let totalWidth = 0;

        for (const char of text) {
            // Get character width
            const charSize = cv.getTextSize(char, font, fontScale, thickness);
            const charWidth = charSize.width;

            // Draw stroke first
            if (strokeRgba && strokeWidth) {
                const strokeThickness = thickness + Math.floor(strokeWidth * 2);
                if (strokeRgba[3] > 0) {
                    cv.putText(
                        canvas.mat,
                        char,
                        new cv.Point(Math.floor(currentX), y),
                        font,
                        fontScale,
                        new cv.Scalar(strokeRgba[0], strokeRgba[1], strokeRgba[2], 255),
                        strokeThickness,
                        cv.LINE_AA
                    );
                }
            }

            // Draw fill
            if (fillRgba && fillRgba[3] > 0) {
                cv.putText(
                    canvas.mat,
                    char,
                    new cv.Point(Math.floor(currentX), y),
                    font,
                    fontScale,
                    new cv.Scalar(fillRgba[0], fillRgba[1], fillRgba[2], 255),
                    thickness,
                    cv.LINE_AA
                );
            }

            currentX += charWidth + letterSpacing;
            totalWidth += charWidth + letterSpacing;
        }

        // Draw text decoration
        TextRenderer._drawDecoration(
            canvas, textDecoration, x, y, Math.floor(totalWidth - letterSpacing),
            textHeight, fillRgba, thickness
        );
    }

    /**
     * Draw text decoration (underline, strikethrough, etc.).
     * @private
     */
    static _drawDecoration(canvas, decoration, x, y, textWidth, textHeight, color, thickness) {
        if (!color || color[3] === 0 || decoration === 'none') {
            return;
        }

        const lineThickness = Math.max(1, Math.floor(thickness / 2));
        const colorScalar = new cv.Scalar(color[0], color[1], color[2], color[3]);

        if (decoration.includes('underline')) {
            const lineY = y + Math.floor(textHeight / 4);
            cv.line(
                canvas.mat,
                new cv.Point(x, lineY),
                new cv.Point(x + textWidth, lineY),
                colorScalar,
                lineThickness,
                cv.LINE_AA
            );
        }

        if (decoration.includes('line-through') || decoration.includes('strikethrough')) {
            const lineY = y - Math.floor(textHeight / 3);
            cv.line(
                canvas.mat,
                new cv.Point(x, lineY),
                new cv.Point(x + textWidth, lineY),
                colorScalar,
                lineThickness,
                cv.LINE_AA
            );
        }

        if (decoration.includes('overline')) {
            const lineY = y - textHeight;
            cv.line(
                canvas.mat,
                new cv.Point(x, lineY),
                new cv.Point(x + textWidth, lineY),
                colorScalar,
                lineThickness,
                cv.LINE_AA
            );
        }
    }

    /**
     * Calculate bounding box for text element.
     * @param {Object} textData - Dictionary with text properties
     * @param {number} scaleX - X scale factor
     * @param {number} scaleY - Y scale factor
     * @returns {number[]} Tuple [x, y, width, height]
     */
    static getTextBounds(textData, scaleX = 1.0, scaleY = 1.0) {
        const text = textData.text;
        const font = TextRenderer.getOpenCVFont(
            textData.font_family || 'sans-serif',
            textData.font_style || 'normal'
        );
        const fontSize = textData.font_size || 16;
        const fontScale = TextRenderer.calculateFontScale(fontSize * Math.min(scaleX, scaleY));

        const fontWeight = textData.font_weight || 'normal';
        const thickness = TextRenderer.FONT_WEIGHT_MAP[String(fontWeight)] || 1;

        const textSize = cv.getTextSize(text, font, fontScale, thickness);
        const textWidth = textSize.width;
        const textHeight = textSize.height;
        const baseline = textSize.baseLine;

        const x = Math.floor(textData.x * scaleX);
        const y = Math.floor(textData.y * scaleY);

        return [x, y - textHeight, textWidth, textHeight + baseline];
    }

    /**
     * Get the position (x, y) of a specific character in the text.
     * @param {Object} textData - Dictionary with text properties
     * @param {number} charIndex - Index of the character (0-based)
     * @param {number} scaleX - X scale factor
     * @param {number} scaleY - Y scale factor
     * @param {number} offsetX - X offset for positioning
     * @param {number} offsetY - Y offset for positioning
     * @returns {number[]|null} Tuple [x, y] of the character position, or null if invalid
     */
    static getCharPosition(textData, charIndex, scaleX = 1.0, scaleY = 1.0, offsetX = 0, offsetY = 0) {
        const text = textData.text;
        if (charIndex < 0 || charIndex > text.length) {
            return null;
        }

        if (charIndex === 0) {
            // Return the starting position
            const x = Math.floor(textData.x * scaleX + offsetX);
            const y = Math.floor(textData.y * scaleY + offsetY);
            return [x, y];
        }

        // Get font settings
        const font = TextRenderer.getOpenCVFont(
            textData.font_family || 'sans-serif',
            textData.font_style || 'normal'
        );
        const fontSize = textData.font_size || 16;
        const fontScale = TextRenderer.calculateFontScale(fontSize * Math.min(scaleX, scaleY));

        const fontWeight = textData.font_weight || 'normal';
        const thickness = TextRenderer.FONT_WEIGHT_MAP[String(fontWeight)] || 1;

        // Calculate base position
        let x = Math.floor(textData.x * scaleX + offsetX);
        let y = Math.floor(textData.y * scaleY + offsetY);

        // Calculate full text size for alignment
        const fullTextSize = cv.getTextSize(text, font, fontScale, thickness);
        const fullWidth = fullTextSize.width;
        const textHeight = fullTextSize.height;

        // Adjust position based on text-anchor
        const textAnchor = textData.text_anchor || 'start';
        if (textAnchor === 'middle') {
            x -= Math.floor(fullWidth / 2);
        } else if (textAnchor === 'end') {
            x -= fullWidth;
        }

        // Adjust baseline
        const dominantBaseline = textData.dominant_baseline || 'auto';
        if (['middle', 'central'].includes(dominantBaseline)) {
            y += Math.floor(textHeight / 2);
        } else if (['hanging', 'text-before-edge'].includes(dominantBaseline)) {
            y += textHeight;
        }

        // Calculate width of text up to charIndex
        const letterSpacing = (textData.letter_spacing || 0) * scaleX;

        if (letterSpacing > 0) {
            // Calculate width character by character
            let currentX = x;
            for (let i = 0; i < Math.min(charIndex, text.length); i++) {
                const char = text[i];
                const charSize = cv.getTextSize(char, font, fontScale, thickness);
                currentX += charSize.width + letterSpacing;
            }
            x = Math.floor(currentX);
        } else {
            // Calculate width of substring
            const substring = text.substring(0, charIndex);
            const subSize = cv.getTextSize(substring, font, fontScale, thickness);
            x += subSize.width;
        }

        return [x, y];
    }
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { TextRenderer };
}

if (typeof window !== 'undefined') {
    window.TextRenderer = TextRenderer;
}
